pkg/controller/build/reconciler.go

```diff
@@ -791,6 +911,150 @@ func (b *buildReconciler) deleteMachineOSBuild(ctx context.Context, mosb *mcfgv1
+	}
+	if secret.Type != corev1.SecretTypeDockerConfigJson {
+		return "", fmt.Errorf("rendered push secret for MachineOSConfig %q is not of type %q, cannot delete image", moscName, corev1.SecretTypeDockerConfigJson)
+	}
+	var authConfig map[string]interface{}
+	data, ok := secret.Data[corev1.DockerConfigJsonKey]
+	if !ok {
+		return "", fmt.Errorf("rendered push secret for MachineOSConfig %q does not have key %q, cannot delete image", moscName, corev1.DockerConfigJsonKey)
+	}
+	if err := json.Unmarshal(data, &authConfig); err != nil {
+		return "", fmt.Errorf("could not unmarshal rendered push secret for MachineOSConfig %q: %w, cannot delete image", moscName, err)
+	}
+	// Create a temp auth.json file
+	authFile, err := os.CreateTemp("", "auth-*.json")
+	if err != nil {
+		return "", fmt.Errorf("could not create temp file for rendered push secret for MachineOSConfig %q: %w, cannot delete image", moscName, err)
+	}
+	if err := os.WriteFile(authFile.Name(), data, daemon.DefaultDirectoryPermissions); err != nil {
+		return "", fmt.Errorf("could not write temp auth file for rendered push secret for MachineOSConfig %q: %w, cannot delete image", moscName, err)
+	}
+	return authFile.Name(), nil
+}
+
+func (b *buildReconciler) deleteMOSBImage(mosb *mcfgv1.MachineOSBuild, moscName string) error {
+	moscExists := true
+	_, err := b.listers.machineOSConfigLister.Get(moscName)
+	if k8serrors.IsNotFound(err) {
+		moscExists = false
+	} else if err != nil {
+		return fmt.Errorf("could not get MachineOSConfig for MachineOSBuild %q: %w", mosb.Name, err)
+	}
+
+	if moscExists {
+		pool, err := b.listers.machineConfigPoolLister.Get(mosb.ObjectMeta.Labels[constants.TargetMachineConfigPoolLabelKey])
+		if err != nil {
+			return fmt.Errorf("could not get MachineConfigPool from MachineOSBuild %q: %w", mosb.Name, err)
+		}
+
+		nodes, err := helpers.GetNodesForPool(b.listers.machineConfigPoolLister, b.listers.nodeLister, pool)
+		if err != nil {
+			return fmt.Errorf("could not get nodes for MachineConfigPool %q: %w", pool.Name, err)
+		}
+
+		for _, node := range nodes {
+			if node.GetAnnotations()[daemonconstants.CurrentImageAnnotationKey] == string(mosb.Status.DigestedImagePushSpec) ||
+				node.GetAnnotations()[daemonconstants.DesiredImageAnnotationKey] == string(mosb.Status.DigestedImagePushSpec) {
+				// the image we are trying to delete is currently on a node or desired by a node
+				klog.Warningf("Image %s is currently applied on a node or desired by a node, will not delete", string(mosb.Status.DigestedImagePushSpec))
+				return nil
+			}
+		}
+	}
+
+	image := string(mosb.Spec.RenderedImagePushSpec)
+	isOpenShiftRegistry, err := b.isOpenShiftRegistry(image)
+	if err != nil {
+		return err
+	}
+	if isOpenShiftRegistry {
+		klog.Infof("Deleting image %s from internal registry for MachineOSBuild %s", image, mosb.Name)
+		// Use the openshift API to delete the image
+		ns, img, err := extractNSAndNameWithTag(image)
```

Comment:
praise: I imagine this must've been a little tricky. But this should allow one to use an imagestream in any namespace. Nicely done!